diag_manager/fms_diag_object.F90

```diff
@@ -39,7 +41,11 @@ module fms_diag_object_mod
 !TODO: Remove FMS prefix from variables in this type
   class(fmsDiagFileContainer_type), allocatable :: FMS_diag_files (:) !< array of diag files
   class(fmsDiagField_type), allocatable :: FMS_diag_fields(:) !< Array of diag fields
+  class(fmsDiagAxis_type), allocatable :: diag_axis(:) !< Array of diag_axis
+  class(fmsDiagSubAxis_type), allocatable :: diag_sub_axis(:) !< Array of diag_sub_axis
   integer, private :: registered_variables !< Number of registered variables
+  integer, private :: registered_axis !< Number of registered axis
+  integer, private :: registered_subaxis !< Number of registered subAxis
```

Comment:
I don't think that is necessary. 

The subaxis needs information about a parent axis (i.e the units, cart_name, data, etc ...) so if we do
```
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type
```
We will have to copy all the info from the parent axis to the sub axis object.
Rather than doing that I just set up a parent_axis_id variable in the sub axis object that we can just to get all the information we need. 

I suppose we have something like:

```F90
type :: fmsDiagAxisContainer_type                                                                                       
  class(fmsDiagAxis_type), allocatable :: axis                                                                          

! PROCEDURES: diag_parent_axis_init
! PROCEDURES: diag_sub_axis_init
end type 

type :: fmsDiagAxis_type                                                                                                
  integer :: axis_id                                                                                                    
! PROCEDURES :: get_axis_length, write_axis_data, write_axis_metadata                                                   
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagParentAxis_type                                                               
  integer :: cart !< axis_data, axis_name, etc etc                                                                      
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type                                                                  
  integer :: parent_axis_id                                                                                             
  !< other subaxis stuff                                                                                                
end type 
```

The diag_axis_init will allocate `axis` to be fmsDiagParentAxis_type and add all the info inside of it
Whenever we need a subaxis we can allocate `axis` to be fmsDiagSubAxis_type and fill it in. 

What do you think?

Reply:
I think this is a great idea, but the naming is wrong.  `fmsDiagParentAxis_type` is not a parent, it's a child.  Maybe something like:
```fortran
type :: fmsDiagAxisContainer_type                                                                                       
  class(fmsDiagAxis_type), allocatable :: axis                                                                          

! PROCEDURES: diag_parent_axis_init
! PROCEDURES: diag_sub_axis_init
end type 

type :: fmsDiagAxis_type                                                                                                
  integer :: axis_id                                                                                                    
! PROCEDURES :: get_axis_length, write_axis_data, write_axis_metadata                                                   
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagFullAxis_type                                                               
  integer :: cart !< axis_data, axis_name, etc etc                                                                      
end type                                                                                                                
                                                                                                                        
type, extends(fmsDiagAxis_type) :: fmsDiagSubAxis_type                                                                  
  integer :: parent_axis_id                                                                                             
  !< other subaxis stuff                                                                                                
end type 
```
Now it differentiates between a `full` axis and a `sub` axis. Don't feel like you have to go with this, it's just a suggestion.  But you should definitely not name that type `fmsDiagParentAxis_type`

Reply:
done in https://github.com/NOAA-GFDL/FMS/pull/1024/commits/f09879e97c7fef3c5900880483d43b293d410e31